src/i_oalsound.c

```diff
@@ -0,0 +1,662 @@
+    {
+        printf(" Failed to find resampler: '%s'.\n", resampler_name);
+        return;
+    }
+
+    for (i = 0; i < MAX_CHANNELS; i++)
+    {
+        alSourcei(sources[i], AL_SOURCE_RESAMPLER_SOFT, def_resampler);
+    }
+/*
+    printf(" Using '%s' resampler.\n",
+           alGetStringiSOFT(AL_RESAMPLER_NAME_SOFT, def_resampler));
+*/
+}
+
+void I_OAL_ResetSource2D(int channel)
+{
+    if (!oal)
+    {
+        return;
+    }
+
+    // Don't reset while sound is playing.
+    if (oal->active[channel])
+    {
+        return;
+    }
+
+    if (oal->SOFT_source_spatialize)
+    {
+        alSourcei(oal->sources[channel], AL_SOURCE_SPATIALIZE_SOFT, AL_FALSE);
+    }
+
+    if (oal->EXT_EFX)
+    {
+        alSourcef(oal->sources[channel], AL_AIR_ABSORPTION_FACTOR, 0.0f);
+    }
+
+    if (oal->EXT_SOURCE_RADIUS)
+    {
+        alSourcef(oal->sources[channel], AL_SOURCE_RADIUS, 0.0f);
+    }
+
+    alSource3f(oal->sources[channel], AL_POSITION, 0.0f, 0.0f, 0.0f);
+    alSource3f(oal->sources[channel], AL_VELOCITY, 0.0f, 0.0f, 0.0f);
+    alSource3i(oal->sources[channel], AL_DIRECTION, 0, 0, 0);
+
+    alSourcei(oal->sources[channel], AL_ROLLOFF_FACTOR, 0);
+    alSourcei(oal->sources[channel], AL_SOURCE_RELATIVE, AL_TRUE);
+}
+
+void I_OAL_ResetSource3D(int channel, boolean point_source)
+{
+    if (!oal)
+    {
+        return;
+    }
+
+    // Don't reset while sound is playing.
+    if (oal->active[channel])
+    {
+        return;
+    }
+
+    if (oal->SOFT_source_spatialize)
+    {
+        alSourcei(oal->sources[channel], AL_SOURCE_SPATIALIZE_SOFT, AL_TRUE);
+    }
+
+    if (oal->EXT_EFX)
+    {
+        alSourcef(oal->sources[channel], AL_AIR_ABSORPTION_FACTOR, oal->absorption);
+    }
+
+    if (oal->EXT_SOURCE_RADIUS)
+    {
+        alSourcef(oal->sources[channel], AL_SOURCE_RADIUS,
+                  point_source ? 0.0f : OAL_SOURCE_RADIUS);
+    }
+
+    alSourcei(oal->sources[channel], AL_ROLLOFF_FACTOR, OAL_ROLLOFF_FACTOR);
+    alSourcei(oal->sources[channel], AL_SOURCE_RELATIVE, AL_FALSE);
+}
+
+void I_OAL_AdjustSource3D(int channel, const ALfloat *position,
+                          const ALfloat *velocity, const ALfloat *direction)
+{
+    if (!oal)
+    {
+        return;
+    }
+
+    alSourcefv(oal->sources[channel], AL_POSITION, position);
+    alSourcefv(oal->sources[channel], AL_VELOCITY, velocity);
+    alSourcefv(oal->sources[channel], AL_DIRECTION, direction);
+}
+
+void I_OAL_AdjustListener3D(const ALfloat *position, const ALfloat *velocity,
+                            const ALfloat *orientation)
+{
+    if (!oal)
+    {
+        return;
+    }
+
+    alListenerfv(AL_POSITION, position);
+    alListenerfv(AL_VELOCITY, velocity);
+    alListenerfv(AL_ORIENTATION, orientation);
+}
+
+void I_OAL_UpdateUserSoundSettings(void)
+{
+    if (!oal)
+    {
+        return;
+    }
+
+    // Source parameters.
+    SetResampler(oal->sources);
+    oal->absorption = (ALfloat)snd_absorption / 2.0f;
```

Comment:
Are these values ever reset e.g. when switching from "OpenAL 3D" mode to "normal" mode?

Reply:
No, I looked at the OpenAL Soft code ([here](https://github.com/kcat/openal-soft/blob/e714c8fbca7bc4a03be308d43d6d6cd5e57914bc/alc/alc.cpp#L3325-L3347) and [here](https://github.com/kcat/openal-soft/blob/e714c8fbca7bc4a03be308d43d6d6cd5e57914bc/alc/alc.cpp#L1783-L1822)) and then did some extensive logging of values before initialization, after initialization, and then after _reinitialization_. OpenAL Soft only changes the values defined by the attributes list which is very convenient. This is also why the music module is unaffected by `alcResetDeviceSOFT()`). So that makes the line here unnecessary: https://github.com/fabiangreffrath/woof/pull/1118/commits/4fc925a64ee4b6c23f27fd65f3fbbb0418e62f1b

Reply:
Actually, I want to keep `ResetParams();` because switching from, say, 3D to 2D means the sources and listener should be updated accordingly.

Reply:
So, do I read from this that air absorption and Doppler effect are also applied in "normal" mode?

Reply:
It's relying on the side effects other variables but that's not good practice so I should explicitly disable them instead.

Reply:
Fixed by https://github.com/fabiangreffrath/woof/pull/1118/commits/e1d2aa93881c625628b0e9d010498a6762bbdaa4